### ISSUE TITLE
Roslyn incorrectly reports local functions as unused if they are used only under nameof as default parameter values

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#16821]
```cs
using System;
public class C 
{
    public void M() 
    {
        void Local() {} // incorrectly reported as unused
        void Local2(string s = nameof(Local)) => Console.WriteLine(s);
        Local2();
    }
}
```

**Comment by khyperia** [comment]
Update while exploring this issue to fix it:

The problem is more general than originally reported as. All default values of parameters in local functions are not marked as used:

```
int x; // Warning: unused
void Foo1(string s = nameof(x)) { }

const int y = 2; // Warning: unused
void Foo2(int v = y) { }
```

This is because `SourceComplexParameterSymbol` does not hold on to the bound expression for the default value. It binds the expression, extracts its constant value, and immediately drops the bound expression. This code is in `SourceComplexParameterSymbol.MakeDefaultExpression`. I plan on storing the bound expression and updating the visitors to visit it, which should fix the use analyzer.

**Comment by jcouv** [comment]
@agocke @khyperia Given the current 15.5 bar, I think we should move this to next release. Let me know if you think otherwise. Thanks

**Comment by jaredpar** [comment]
Given there is a PR out that fixes this issue, just didn't meet an earlier bar, keeping it in 15.7.

**Comment by CyrusNajmabadi** [comment]
I don't repro this:

<img width="1284" height="271" alt="Image" src="https://github.com/user-attachments/assets/c7d78bb3-f6f7-4a6f-bf33-0720a4910f6c" />

Build produces no messages about this being unused either.
